compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/inlineclasses/InlineClassAbi.kt

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.backend.jvm.lower.inlineclasses
+
+import org.jetbrains.kotlin.backend.jvm.ir.erasedUpperBound
+import org.jetbrains.kotlin.ir.declarations.*
+import org.jetbrains.kotlin.ir.types.IrType
+import org.jetbrains.kotlin.ir.types.classOrNull
+import org.jetbrains.kotlin.ir.types.isPrimitiveType
+import org.jetbrains.kotlin.ir.types.makeNullable
+import org.jetbrains.kotlin.ir.util.*
+import org.jetbrains.kotlin.load.java.JvmAbi
+import org.jetbrains.kotlin.name.Name
+import java.security.MessageDigest
+import java.util.*
+
+/**
+ * Replace inline classes by their underlying types.
+ */
+fun IrType.unbox() = InlineClassAbi.unboxType(this) ?: this
+
+/**
+ * Check if a function can be replaced.
+ */
+val IrFunction.isBoxed: Boolean
+    get() = explicitParameters.any { it.type.erasedUpperBound.isInline } || (this is IrConstructor && constructedClass.isInline)
+
+object InlineClassAbi {
+    /**
+     * Unwraps inline class types to their underlying representation.
+     * Returns null if the type cannot be unboxed.
+     */
+    internal fun unboxType(type: IrType): IrType? {
+        val klass = type.classOrNull?.owner ?: return null
+        if (!klass.isInline) return null
+
+        val underlyingType = getUnderlyingType(klass).unbox()
+        if (!type.isNullable())
+            return underlyingType
+        if (underlyingType.isNullable() || underlyingType.isPrimitiveType())
+            return null
+        return underlyingType.makeNullable()
+    }
+
+    /**
+     * Get the underlying type of an inline class based on the single argument to its
+     * primary constructor. This is what the current jvm backend does.
+     *
+     * Looking for a backing field does not work for built-in inline classes (UInt, etc.),
```

Comment:
I'd replace "built-in inline classes (UInt, etc.)" with "unsigned types" to avoid confusion with types that are named built-in (`KotlinBuiltIns`) in the sense that any program has them in dependencies (a program can successfully compile without `UInt` in dependencies)

Reply:
Done